### PR TITLE
chore: remove bumpversion for setup.py and __version__.py [INFENG-704]

### PR DESCRIPTION
**Pull request by davidfluck-hpe**
## Ticket

INFENG-704

## Description

This changeset is the first of some work to rewrite the Determined release process, particularly focusing on being able to tag releases without any additional commits from tools like `bumpversion`. To that end, this changeset removes `bumpversion` dependencies from all `setup.py` files and `__version__.py` files targeted by `.bumpversion.cfg`, and replaces it with a call to Python's `importlib.metadata.version()` to fetch the package version at runtime. Specifically:

```
[bumpversion:glob:*/setup.py]

[bumpversion:glob:*/*/__version__.py]
```

Note that this pull request is against a feature branch, `INFENG-382-release-redesign`. My hope is to use this target branch as a long-running feature branch that I will merge into and keep up-to-date with `main` as I go, until it's ready to be merged into `main`. This seemed a bit easier than attempting to piecemeal merge things into `main`, which would probably necessitate a lot more scaffolding to keep things appropriately split.

Before, the `model-hub` and `determined` Python packages used setuptools to set package attributes via the setup() function. I have migrated most of those to use pyproject.toml instead, which now supports many of the setup() options that we need. (Things that do not seem supported remain in setup().)

I have added a build-time dependency on setuptools-scm, which is a setuptools plugin to dynamically fetch information from source control. This allows setuptools to fetch the version from a git tag, and use it throughout the build process.

I have bumped and pinned the setuptools version to 70.0.0, to support setuptools-scm.

Finally, I removed `__version__.py`, which had the Determined version hardcoded in, and replaced it with a call to `importlib.metadata` in `__init__.py`, which sets the `__version__` variable there instead. Any `determined.__version__` reference will still work, it will just grab the metadata from the Python package instead of a separate Python file on disk. This obviates the need to do any version modification at build time, because `setuptools-scm` fetches it from git instead and puts it into the package metadata.

## Testing

The best way to test this is to test a local build for each Python package. To test `determined`:

   1. Ensure you are in your determined venv, if using a virtual env.
   1. Create a phony git tag for testing: `git tag 0.33.0.dev0`. You won't push this, so it can be anything that conforms to semver.
   1. From `determined/`, install `requirements.txt` to fetch updated setuptools and setuptools-scm: `pip install -r requirements.txt`.
   1. From `determined/harness`, run: `python -m build -nxw`.
   1. If the build completed successfully, you should have a `dist/` directory with `determined-0.33.0.dev0-py3-none-any.whl` in it.
   1. Install the built package locally: `pip install dist/determined-0.33.0.dev0-py3-none-any.whl`.
   1. Run `pip show determined` to inspect the package information. The version should match our tag version:
      ```
      Name: determined
      Version: 0.33.0.dev0
      Summary: Determined AI: The fastest and easiest way to build deep learning models.
      Home-page:
      Author:
      Author-email: Determined AI <ai-open-source@hpe.com>
      License:
      Location: /Users/fluckdav/.venv/determined/lib/python3.8/site-packages
      Requires: analytics-python, appdirs, argcomplete, azure-core, azure-storage-blob, boto3, certifi, docker, filelock, gitpython, google-api-python-client, google-cloud-storage, lomond, matplotlib, numpy, packaging, paramiko, pathspec, psutil, pyOpenSSL, python-dateutil, pytz, pyzmq, requests, ruamel.yaml, tabulate, termcolor, tqdm
      Required-by: model-hub
	  ```

	  Note that despite lacking homepage, author, and license fields, this information is correct in pyproject.toml, as well as in the resulting package METADATA. The fields will be correct on PyPI; this is purely outdated behavior in `pip show`. See [this GitHub thread](https://github.com/pypa/packaging-problems/issues/670) for more information.

   1. Run `make clean` in `determined/harness` to clean up build artifacts.
   
Testing `model-hub` is similar:

   1. From `determined/model_hub`, run: `python -m build -nxw`.
   1. If the build completed successfully, you should have a `dist/` directory with `model_hub-determined-0.33.0.dev0-py3-none-any.whl` in it.
   1. Install the built package locally: `pip install dist/model_hub-determined-0.33.0.dev0-py3-none-any.whl`.
   1. Run `pip show model-hub` to inspect the package information. The version should match our tag version:
      ```
      Name: model-hub
      Version: 0.33.0.dev0
      Summary: Model Hub for Determined Deep Learning Training Platform
      Home-page:
      Author:
      Author-email: Determined AI <ai-open-source@hpe.com>
      License:
      Location: /Users/fluckdav/.venv/determined/lib/python3.8/site-packages
      Requires: determined
      Required-by:
	  ```
   1. Delete the git tag you created earlier: `git tag -d v0.33.0.dev0`.
   1. Uninstall the local libraries: `pip uninstall model_hub determined`.

Finally, we need to test the actual metadata fetching functionality. Run `det --version`, and you should see the tag you set: `det 0.33.0.dev0`.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/pyproject.toml [hunk 1]
```diff
@@ -1,4 +1,86 @@
...
+[project]
+name = "determined"
+description = "Determined AI: The fastest and easiest way to build deep learning models."
+requires-python = ">=3.8"
+
+authors = [
+  { name = "Determined AI", email = "ai-open-source@hpe.com" },
+]
+
+dependencies = [
```
Comment (dannysauer) [inline harness/pyproject.toml:15]: Does pip support reading dependencies from here yet, or is including requirements as a dynamic section from file finally out of beta?  I could read the docs, but I'd rather make a comment on a PR. :D

File: harness/pyproject.toml [hunk 2]
```diff
@@ -1,4 +1,86 @@
...
 [tool.black]
 line-length = 100
 exclude = '(_gen.py|determined/_swagger/client/*)'
 
+[tool.setuptools_scm]
+# Path relative_to pyproject.toml. Necessary for correct git metadata parsing.
+root = "../"
+
+# Don't add the extra commit hash. PyPI complains, and it's also just noisy.
+local_scheme = "no-local-version"
```
Comment (dannysauer) [inline harness/pyproject.toml:76]: Will this potentially cause issues with generating artifacts for per-PR builds?

File: harness/pyproject.toml (hunk 1 above)
Comment (davidfluck-hpe) [reply harness/pyproject.toml:15]: I.e. in lieu of of `requirements.txt`?

File: harness/pyproject.toml (hunk 2 above)
Comment (davidfluck-hpe) [reply harness/pyproject.toml:76]: I don't think so. The current Python package build step uses whatever version is hardcoded on disk, and that doesn't change between commits in a PR, either.

File: harness/pyproject.toml (hunk 1 above)
Comment (dannysauer) [reply harness/pyproject.toml:15]: Yup

**Review by dannysauer — APPROVED** [review]
I added comments but forgot to approve. Whoops.

This looks good.

**Review by jgongd — APPROVED** [review]
LGTM

File: model_hub/pyproject.toml [hunk 3]
```diff
@@ -1,2 +1,43 @@
...
+  { name = "Determined AI", email = "ai-open-source@hpe.com" },
+]
+
+dependencies = [
+  "determined>=0.13.11",  # We require custom reducers for PyTorchTrial.
+]
+
+classifiers = [
+  "License :: OSI Approved :: Apache Software License",
+]
```
Comment (jgongd) [inline model_hub/pyproject.toml:21]: Include license = {text = "Apache License 2.0"}?

Comment (davidfluck-hpe) [reply model_hub/pyproject.toml:21]: Sorry it took me so long to respond. `classifiers` is the current, PyPI-friendly method of specifying a license; as far as I can tell, it's only things like `pip show` that depend on that key now. I'm hesitant to add it back in if it's not necessary, but I will if you tell me to.

Comment (davidfluck-hpe) [reply model_hub/pyproject.toml:21]: Merging for now to get some stuff over the line and clean up branches, but let me know and I can open another PR to add the license string back.

File: harness/pyproject.toml (hunk 1 above)
Comment (davidfluck-hpe) [reply harness/pyproject.toml:15]: It does not, as far as I understand.

**Pull request merged**